### ISSUE TITLE
[MPS] torch.cumprod, cumsum, and sort SIGABRT on MPS for ndim > 4 with left-axis dim

### ISSUE THREAD

**Issue by christianahmed905** [pytorch/pytorch#187105]
### 🐛 Describe the bug

"""
torch.cumprod / torch.cumsum / torch.sort — SIGABRT (process crash) on MPS
for tensors with ndim > 4

Three PyTorch ops crash the Python process with SIGABRT (exit code -6) on MPS
when the tensor rank is high enough that the sorted/scanned axis maps to
position >= 4 from the right.  No Python exception is raised; the process is
killed and cannot be recovered.

Crash origins:
  MPSNDArrayScan.mm:251: failed assertion `(null)' Axis = 4.
    This class only supports axis = 0, 1, 2, 3
    (torch.cumprod, torch.cumsum)

  MPSNDArraySort.mm:252: failed assertion `(null)' Axis = 4.
    This class only supports axis = 0, 1, 2, 3
    (torch.sort)

Root cause: the MPS kernels MPSNDArrayScan and MPSNDArraySort right-align
dimensions and only handle the rightmost 4 axes.  PyTorch does not validate
this constraint before dispatching to MPS, so tensors with ndim > 4 and a
dim value in the "left" portion trigger a fatal assertion inside the Metal
runtime.

General crash rule:  crash when  dim < ndim - 4  (i.e. axis from right >= 4)

Failing combinations (exhaustive, torch.ones):
  op              ndim  dim   result
  cumprod/cumsum   5     0    CRASH  (axis from right = 4)
  cumprod/cumsum   6     0    CRASH  (axis from right = 5)
  cumprod/cumsum   6     1    CRASH  (axis from right = 4)
  sort             5     0    CRASH  (axis from right = 4)
  cumprod/cumsum   5    1-4   ok
  cumprod/cumsum   6    2-5   ok
  sort             5    1-4   ok
  any op          ≤4   any    ok

Crash statistics (Specialized generator, mps):
  torch.cumprod / torch.cumsum — SIGABRT on ndim=5+ with dim=0 (or dim=1 for ndim=6)
  torch.sort                   — 1 crash (hash [PROC], signal=6)
                                 example: shape=(64,16,8,8,16) dim=0 dtype=bfloat16

torch.argsort is also likely affected via the same MPSNDArraySort kernel.

Expected: raise RuntimeError with a clear message (e.g. "MPS supports at most
          4 dimensions for this operation"), or fall back to CPU automatically.
Actual:   SIGABRT kills the Python process — not catchable with try/except.

Affected ops:    torch.cumprod, torch.cumsum (MPSNDArrayScan)
                 torch.sort (and likely torch.argsort) (MPSNDArraySort)
Affected device: mps (Apple Silicon)
Not affected:    ndim <= 4 on MPS; any ndim on CPU or CUDA
Versions:        PyTorch 2.12.0, macOS 26.4.1 arm64 (Apple Silicon MPS)
"""

import torch

print(f"torch version: {torch.__version__}")
assert torch.backends.mps.is_available(), "MPS not available — run on Apple Silicon"

device = torch.device("mps")

# ── Safe inputs (ndim ≤ 4) ────────────────────────────────────────────────────
x4 = torch.ones(2, 2, 2, 2, device=device)
print(f"[ok] cumprod ndim=4 dim=0: {torch.cumprod(x4, dim=0).shape}")
print(f"[ok] cumsum  ndim=4 dim=0: {torch.cumsum(x4, dim=0).shape}")
x4s = torch.ones(4, 4, 4, 4, dtype=torch.bfloat16, device=device)
print(f"[ok] sort    ndim=4 dim=0: {torch.sort(x4s, dim=0).values.shape}")

# ── Safe high-rank inputs (rightmost 4 dims) ──────────────────────────────────
x5 = torch.ones(2, 2, 2, 2, 2, device=device)
print(f"[ok] cumprod ndim=5 dim=1 (axis from right=3): {torch.cumprod(x5, dim=1).shape}")
x5s = torch.ones(64, 16, 8, 8, 16, dtype=torch.bfloat16, device=device)
print(f"[ok] sort    ndim=5 dim=1 (axis from right=3): {torch.sort(x5s, dim=1).values.shape}")

print()
print("Choosing which crash to trigger (pick one — each kills the process):")
print("  Uncomment exactly one block below and re-run.\n")

# WARNING: each block below kills the process with SIGABRT.
# Uncomment ONE at a time.

# -- cumprod/cumsum ndim=5 dim=0 --
# print("ndim=5 dim=0 cumprod → SIGABRT")
# torch.cumprod(x5, dim=0)

# -- sort ndim=5 dim=0 --
# print("ndim=5 dim=0 sort → SIGABRT")
# torch.sort(x5s, dim=0, stable=True)

print("(no crash triggered — uncomment a block above to reproduce)")

### Versions

Collecting environment information...
PyTorch version: 2.12.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 26.5 (arm64)
GCC version: Could not collect
Clang version: 21.0.0 (clang-2100.1.1.101)
CMake version: Could not collect
Libc version: N/A

Python version: 3.14.5 (main, May 10 2026, 19:20:57) [Clang 22.1.3 ] (64-bit runtime)
Python platform: macOS-26.5-arm64-arm-64bit-Mach-O
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Apple M5

Versions of relevant libraries:
[pip3] numpy==2.4.6
[pip3] torch==2.12.0
[conda] Could not collect

cc @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk @Isalia20

**Comment by Isalia20** [comment]
This is fixed on main/nightly
